src/util/threadpool.rs

```diff
@@ -211,21 +115,28 @@ where
 
         ThreadPool {
             task_pool: task_pool,
+            stop_flag: stop_flag,
             threads: threads,
             task_count: task_count,
-            sender: sender,
         }
     }
 
-    pub fn execute<F>(&mut self, gid: T, job: F)
+    pub fn execute<F>(&mut self, job: F)
     where
-        F: FnOnce() + Send + 'static,
+        F: FnOnce(&mut Ctx) + Send + 'static,
+        Ctx: Context,
     {
-        let task = Task::new(gid, job);
-        self.sender.send(task).unwrap();
-        self.task_count.fetch_add(1, AtomicOrdering::SeqCst);
-        let &(_, ref cvar) = &*self.task_pool;
+        if self.stop_flag.load(AtomicOrdering::SeqCst) {
+            return;
+        }
+        let task = Task::new(job);
+        let &(ref lock, ref cvar) = &*self.task_pool;
+        {
+            let mut queue = lock.lock().unwrap();
```

Comment:
Although it should be ok to notify thread without holding the mutex, I prefer to follow the good pattern and don't do unnecessary optimization.